ddprof-lib/src/main/java/com/datadoghq/profiler/BufferWriter.java

```diff
@@ -0,0 +1,181 @@
+package com.datadoghq.profiler;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Version-agnostic wrapper for direct ByteBuffer memory operations with explicit memory ordering guarantees.
+ *
+ * <p>This class provides low-level memory access operations on direct ByteBuffers with precise control
+ * over memory ordering semantics. It abstracts differences between Java 8 (using sun.misc.Unsafe) and
+ * Java 9+ (using VarHandle APIs) to provide consistent behavior across JVM versions.
+ *
+ * <p>The class supports two types of memory ordering:
+ * <ul>
+ *   <li><b>Ordered writes:</b> Prevents reordering with prior writes but allows subsequent operations
+ *       to be reordered before this write. More efficient than volatile writes.</li>
+ *   <li><b>Volatile writes:</b> Full memory barrier - prevents reordering with both prior and subsequent
+ *       operations. Ensures immediate visibility to all threads.</li>
+ * </ul>
+ *
+ * <p>This is primarily used by the profiler for writing thread-local context data (span IDs, checksums)
+ * to direct ByteBuffers that are shared between Java and native code via JNI.
+ *
+ * <p><b>Thread Safety:</b> This class is thread-safe. Individual write operations provide their own
+ * memory ordering guarantees as documented.
+ */
+public final class BufferWriter {
+    /**
+     * Service Provider Interface for version-specific buffer write implementations.
+     *
+     * <p>Implementations of this interface provide the actual low-level memory access operations:
+     * <ul>
+     *   <li>{@code BufferWriter8} - Java 8 implementation using sun.misc.Unsafe</li>
+     *   <li>{@code BufferWriter9} - Java 9+ implementation using VarHandle</li>
+     * </ul>
+     *
+     * <p>The implementation is selected automatically based on the runtime Java version.
+     */
+    public interface Impl {
+        /**
+         * Writes a long value to the buffer at the specified offset with ordered write semantics.
+         *
+         * <p>Ordered write guarantees that this write will not be reordered with respect to prior writes,
+         * but subsequent operations may be reordered before this write. This is more efficient than a
+         * volatile write when full bidirectional ordering is not required.
+         *
+         * @param buffer the direct ByteBuffer to write to
+         * @param offset the offset in bytes from the buffer's base address
+         * @param value the long value to write
+         * @return the written value
+         */
+        long writeLong(ByteBuffer buffer, int offset, long value);
```

Comment:
Maybe rename to writeLongOrdered() to make it clear that this is not just a plain write?
Also, have you verified that calls to the implementation are properly inlined? Otherwise you would lose the advantage of ordered vs volatile writes.